README.md
A set helpers to make it easy to use [Shelf][shelf] on App Engine.

[shelf]: https://pub.dartlang.org/packages/shelf

## Running the Example

Example code for this package does not follow [Dart conventions][example]. The
package is structured so it can be [run directly using gcloud][run].

**Using `pub build`**

The easiest way to run the sample is to run `pub build` before you execute
`gcloud preview app run app.yaml`. If you change the content of the `web`
Directory, you will have to rerun `pub build`.

**Using `pub serve`**

If you'd like to use `pub serve` during development, follow the instructions
[here][serve]. Note: you will still need to run `pub build` before you deploy.

[example]: https://www.dartlang.org/tools/pub/package-layout.html#examples
[run]: https://www.dartlang.org/cloud/run.html#run-the-app-using-app-engine
[serve]: https://www.dartlang.org/cloud/client-server/#get-the-clientserver-code-and-run-it-locally


CHANGELOG.md
## 0.3.0-alpha

* Starting on `flex` support with `pkg/appengine` v`0.4.0-flex.alpha.0+4`.

## 0.2.3

* Added `port` argument to `serve` function. Requires `appengine >= 0.3.1`.

## 0.2.2+1

* Better handling of default document serving.

## 0.2.2

* Support latest release of `appengine` package.

## 0.2.1

* Made `DirectoryIndexServeMode` an enum.

* Support the latest version of `shelf` package.

* Require Dart 1.9 or greater.

## 0.2.0+1

* Fixing an issue causing the `DirectoryIndexServeMode.SERVE` mode to have no
  effect.

## 0.2.0

* Made `assetHandler` a function.

* Added the `directoryIndexServeMode` named parameter to the `assetHandler`
  method to enable auto-serving or redirecting to `index.html` files.
  Allow changing the default index files name to serve with `indexFileName`.

## 0.1.1+2

* Formatted the code.

* Updated example code to run on the latest configuration.

## 0.1.1+1

* Added logging for asset errors.

## 0.1.1

* First public release.


test/test_app_run.sh
#!/usr/bin/env bash
exec gcloud preview app run \
  test/dispatch.yaml \
  test/module-serve/module-serve.yaml \
  test/module-redirect/module-redirect.yaml \
  test/default/default.yaml


test/shelf_appengine_test.dart
// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

import 'dart:async';
import 'dart:collection';
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:path/path.dart' as p;
import 'package:scheduled_test/scheduled_test.dart';
import 'package:scheduled_test/descriptor.dart' as d;

import 'package:shelf_appengine/shelf_appengine.dart';

Stream<String> _getLines(Stream<List<int>> stdout) =>
    stdout.transform(SYSTEM_ENCODING.decoder).transform(const LineSplitter());

const _defaultPubServeHost = 'http://192.168.59.3:7777';
const _pubServeEnvKey = "DART_PUB_SERVE";

Uri _serverUri;

void main() {
  _runModes(false);

  group('with pub serve', () {
    _runModes(true);
  });
}

void _runModes(bool withPubServe) {
  group('default value (same as none)', () {
    _runTest(null, 404, withPubServe);
  });

  group('none', () {
    _runTest(DirectoryIndexServeMode.NONE, 404, withPubServe);
  });

  group('serve', () {
    _runTest(DirectoryIndexServeMode.SERVE, 200, withPubServe);
  });

  group('redirect', () {
    _runTest(DirectoryIndexServeMode.REDIRECT, 302, withPubServe);
  });
}

void _runTest(DirectoryIndexServeMode mode, int dirCode, bool withPubServe) {
  Directory tempDir;
  Process appEngineProcess;
  Process pubServeProcess;

  String workingDir() => p.join(tempDir.path, 'pkg');

  setUp(() {
    // set up the dir
    schedule(() async {
      var dir = await Directory.systemTemp.createTemp('shelf_appengine.test.');

      tempDir = dir;
      d.defaultRoot = tempDir.path;
    });

    schedule(() async {
      d.dir('pkg', [
        d.file('Dockerfile', _dockerFile),
        d.file('app.yaml', _getAppYaml(withPubServe)),
        d.file('pubspec.yaml', _pubspecYaml),
        d.dir('bin', [d.file('server.dart', _getServerCode(mode))]),
        d.dir('build', [
          d.dir('web', [d.file('index.html', _indexHtml)])
        ])
      ]).create();
    });

    // pub build
    schedule(() {
      var result = Process.runSync('pub', ['install', '--offline'],
          workingDirectory: workingDir());

      expect(result.exitCode, 0);
    });

    schedule(() async {
      var hostPort = await _getOpenPort();
      var defaultHost = 'localhost:$hostPort';

      var adminPort = await _getOpenPort();
      var adminHost = 'localhost:$adminPort';

      appEngineProcess = await Process.start(
          'gcloud',
          [
            '--project',
            'shelf-appengine-test',
            'preview',
            'app',
            'run',
            'app.yaml',
            '--host',
            defaultHost,
            '--admin-host',
            adminHost
          ],
          workingDirectory: workingDir());

      var waitingForNext = false;

      Completer answerCompleter = new Completer();

      StreamSubscription sub;

      sub = _getLines(appEngineProcess.stderr).listen((line) {
        print(line);
        if (answerCompleter.isCompleted) return;

        if (line.startsWith('ERROR:')) {
          answerCompleter.completeError(line);
          sub.cancel();
          return;
        }

        if (line.contains('New instance for module "default" serving on:')) {
          waitingForNext = true;
        } else if (waitingForNext) {
          answerCompleter.complete(line);

          // NOTE: comment out this cancel to see what's going on
          // sub.cancel();
        }
      });

      _serverUri = Uri.parse(await answerCompleter.future);
    });
    // TODO: wait for the right output?

    // TODO: kill pub server, if we're doing that

    currentSchedule.onComplete.schedule(() async {
      if (appEngineProcess != null) {
        int rootPid = appEngineProcess.pid;

        // NOW! *sigh* go looking for child processes of the just kill process
        var result = Process.runSync('ps', ['-o', 'pid, ppid']);

        if (result.exitCode == 0) {
          //print([result.stdout, result.stderr]);

          var parentTree = new Map<int, Set<int>>();

          for (var match in _pidRegexp.allMatches(result.stdout)) {
            var pid = int.parse(match[1]);
            var ppid = int.parse(match[2]);

            var parentSet = parentTree.putIfAbsent(ppid, () => new Set<int>());
            parentSet.add(pid);
          }

          //print(parentTree);

          var toKill = new LinkedHashSet<int>();
          var toProcess = <int>[rootPid];

          //print("root pid: $rootPid");

          while (toProcess.isNotEmpty) {
            var current = toProcess.removeLast();
            toKill.add(current);

            var kids = parentTree[current];

            if (kids == null) {
              continue;
            }

            toProcess.addAll(kids);
          }

          //print(toKill);

          for (var pid in toKill) {
            //print('killing $pid');
            var output = Process.killPid(pid);
            expect(output, isTrue);
            //print('Success: $output');
          }
        }

        // print("waiting for things to die...");
        await appEngineProcess.exitCode;
        appEngineProcess = null;
      }

      if (pubServeProcess != null) {
        pubServeProcess.kill(ProcessSignal.SIGTERM);

        await pubServeProcess.exitCode;
      }

      d.defaultRoot = null;
      return tempDir.delete(recursive: true);
    });
  });

  test("Request 'dir'", () {
    schedule(() async {
      var target = _serverUri.resolve('/');
      var statusCode = await _getNoRedirect(target);
      expect(statusCode, dirCode);
    });

    schedule(() async {
      var target = _serverUri.resolve('/index.html');
      var statusCode = await _getNoRedirect(target);
      expect(statusCode, 200);
    });
  }, timeout: new Timeout(const Duration(minutes: 3)));
}

Future<int> _getNoRedirect(Uri uri) async {
  var client = new http.IOClient();
  try {
    var request = new http.Request('GET', uri)..followRedirects = false;
    var streamedResponse = await client.send(request);
    return streamedResponse.statusCode;
  } finally {
    client.close();
  }
}

String _getServerCode(DirectoryIndexServeMode mode) {
  String modeText;

  if (mode == null) {
    modeText = '';
  } else {
    switch (mode) {
      case DirectoryIndexServeMode.NONE:
        modeText = 'directoryIndexServeMode: DirectoryIndexServeMode.NONE';
        break;
      case DirectoryIndexServeMode.SERVE:
        modeText = 'directoryIndexServeMode: DirectoryIndexServeMode.SERVE';
        break;
      case DirectoryIndexServeMode.REDIRECT:
        modeText = 'directoryIndexServeMode: DirectoryIndexServeMode.REDIRECT';
        break;
      default:
        throw 'Not supported - $mode';
    }
  }

  return '''
import 'package:shelf_appengine/shelf_appengine.dart';

void main() {
  serve(assetHandler(${modeText}));
}
''';
}

// TODO: should make the shared port thing correct-ish
String _getAppYaml(bool usePubServe) {
  var env = '';

  if (usePubServe) {
    var envValue = Platform.environment[_pubServeEnvKey];

    if (envValue == null) {
      print('$_pubServeEnvKey not set. Using default: $_defaultPubServeHost');
      envValue = _defaultPubServeHost;
    } else {
      print('Using ENV $_pubServeEnvKey - $envValue');
    }

    env = '''
env_variables:
  DART_PUB_SERVE: $envValue
  ''';
  }

  return '''
runtime: custom
vm: true
api_version: 1
${env}''';
}

const _pubspecYaml = r'''
name: _test_pkg
version: 0.2.0
dependencies:
  shelf_appengine: any
''';

const _indexHtml = r'''
<html><body>hello</body></html>''';

const _dockerFile = r'''FROM google/dart-runtime''';

var _pidRegexp = new RegExp(r'(\d+)\s+(\d+)');

/// Returns an open port by creating a temporary Socket
Future<int> _getOpenPort() async {
  ServerSocket socket;

  try {
    socket = await ServerSocket.bind(InternetAddress.LOOPBACK_IP_V4, 0);
  } catch (_) {
    // try again v/ V6 only. Slight possibility that V4 is disabled
    socket = await ServerSocket.bind(InternetAddress.LOOPBACK_IP_V6, 0,
        v6Only: true);
  }

  try {
    return socket.port;
  } finally {
    await socket.close();
  }
}


lib/shelf_appengine.dart
// Copyright (c) 2014, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

/// Provides hosting of `Shelf` handlers via App Engine.
library shelf_appengine;

import 'dart:async';
import 'dart:io' as io;

import 'package:appengine/appengine.dart' as ae;
import 'package:mime/mime.dart' as mime;
import 'package:path/path.dart' as p;
import 'package:shelf/shelf.dart';
import 'package:shelf/shelf_io.dart' as shelf_io;

/// Serves the provided [handler] using [ae.runAppEngine].
Future serve(Handler handler, {Function onError}) {
  return ae.runAppEngine((io.HttpRequest request) {
    shelf_io.handleRequest(request, handler);
  }, onError: onError);
}

/// Modes available to serve index files in directories.
enum DirectoryIndexServeMode {
  /// When a directory URI is requested no special actions are taken. This
  /// usually end up in a 404 Not Found response.
  NONE,

  /// When a directory URI is requested serve the index file directly in the
  /// response.
  SERVE,

  /// When a directory URI is requested redirect to the index file with a
  /// `302 Found` response.
  REDIRECT,

  /// When a directory URI is requested redirect to the index file with a
  /// `303 See Other` response.
  REDIRECT_SEE_OTHER,

  /// When a directory URI is requested redirect to the index file with a
  /// `301 Moved Permanently` response.
  REDIRECT_PERMANENT
}

/// Serves files using [ae.context.assets].
///
/// You can choose how/if index files will be served when a directory URI is
/// requested by setting `directoryIndexServeMode`.
/// [DirectoryIndexServeMode.NONE] is the default. See
/// [DirectoryIndexServeMode] for more options.
/// The default name of the index files to serve can also be changed using
/// `indexFileName`. `index.html` is the default.
//TODO(kevmoo) better docs.
Handler assetHandler(
    {DirectoryIndexServeMode directoryIndexServeMode: DirectoryIndexServeMode.NONE,
    String indexFileName: "index.html"}) => (Request request) async {
  var path = request.url.path;
  var indexPath = p.join(path, indexFileName);

  bool isBarePath = false;
  if (path.isEmpty) {
    isBarePath = request.handlerPath.endsWith('/');
  } else if (path.endsWith('/')) {
    isBarePath = true;
  }

  // If the path requested is a directory root we might serve an index.html
  // file depending on [directoryIndexServeMode].
  if (isBarePath) {
    if (directoryIndexServeMode == DirectoryIndexServeMode.SERVE) {
      path = indexPath;
    } else if (directoryIndexServeMode == DirectoryIndexServeMode.REDIRECT) {
      return new Response.found(indexPath);
    } else if (directoryIndexServeMode ==
        DirectoryIndexServeMode.REDIRECT_SEE_OTHER) {
      return new Response.seeOther(indexPath);
    } else if (directoryIndexServeMode ==
        DirectoryIndexServeMode.REDIRECT_PERMANENT) {
      return new Response.movedPermanently(indexPath);
    }
  }

  // When serving off the file system, the appengine AssetManager just joins
  // the root with the path w/ '+' – need to make sure that's a clean concat
  // TODO(kevmoo) should likely open an issue on this.
  if (!path.startsWith('/')) {
    path = '/' + path;
  }

  try {
    var stream = await ae.context.assets.read(path);

    Map headers;
    var contentType = mime.lookupMimeType(path);
    if (contentType != null) {
      headers = <String, String>{io.HttpHeaders.CONTENT_TYPE: contentType};
    }

    return new Response.ok(stream, headers: headers);
  } catch (err, stack) {
    ae.context.services.logging
        .error('Error getting asset at path $path\n$err\n$stack');
    // TODO(kevmoo): handle only the specific case of an asset not found
    // https://github.com/dart-lang/appengine/issues/7
    if (err is ae.AssetError) {
      return new Response.notFound('not found');
    }
    rethrow;
  }
};
